File: src/org/usfirst/frc/team4322/command/CommandBuilder.java
package org.usfirst.frc.team4322.command;

import java.util.ArrayList;
import java.util.function.Consumer;
import java.util.function.Predicate;

/**
 * Created by nicolasmachado on 4/20/16.
 */
public class CommandBuilder
{
	private Consumer<Command> act = ((c)->{});
	private Consumer<Command> init = ((c)->{});
	private Predicate<Command> cond = ((c)->false);
	ArrayList<Subsystem> subsystems = new ArrayList<>();
	private Consumer<Command> onEnd = (c) -> {};
	private Consumer<Command> onInt = (c) -> {};
	long timeout = 0;
	public static CommandBuilder create()
	{
		return new CommandBuilder();
	}
	public CommandBuilder execute(Consumer<Command> act)
	{
		this.act = act;
		return this;
	}

	public CommandBuilder onInit(Consumer<Command> init)
    {
        this.init = init;
        return this;
    }

	public CommandBuilder runForTime(long millis)
	{
		this.cond = (Command c) -> c.runTimeMillis() < millis;
		return this;
	}
	public CommandBuilder withTimeout(long millis)
	{
		this.timeout = millis;
		return this;
	}
	public CommandBuilder runWhile(Predicate<Command> cond)
	{
		this.cond = cond;
		return this;
	}
	public CommandBuilder onEnd(Consumer<Command> onEnd)
	{
		this.onEnd = onEnd;
		return this;
	}
	public CommandBuilder onInterrupt(Consumer<Command> onInt)
	{
		this.onInt = onInt;
		return this;
	}
	public CommandBuilder require(Subsystem s)
	{
		subsystems.add(s);
		return this;
	}
	public CommandBuilder requires(Subsystem... s)
	{
		for(Subsystem sys : s)
		{
			subsystems.add(sys);
		}
		return this;
	}
	public Command build()
	{
		return new Command(subsystems,timeout) {
			@Override
            protected void initialize()
			{
				init.accept(this);
			}

			@Override
            protected void end()
			{
				onEnd.accept(this);
			}

			@Override
            protected void interrupted()
			{
				onInt.accept(this);
			}

			@Override
            protected boolean isFinished()
			{
				return !cond.test(this);
			}

			@Override
            protected void execute()
			{
				act.accept(this);
			}
		};
	}
}

File: src/org/usfirst/frc/team4322/command/CommandBuilder.kt
package org.usfirst.frc.team4322.command

import java.util.ArrayList
import java.util.function.Consumer
import java.util.function.Predicate

/**
 * Created by nicolasmachado on 4/20/16.
 */
class CommandBuilder {
    private var act : (Command) -> Unit = {}
    private var init : (Command) -> Unit = {}
    private var cond : (Command) -> Boolean = { false }
    internal var subsystems = ArrayList<Subsystem>()
    private var onEnd : (Command) -> Unit = {}
    private var onInt : (Command) -> Unit = {}
    internal var timeout: Long = 0
    fun execute(act: (Command) -> Unit): CommandBuilder {
        this.act = act
        return this
    }

    fun onInit(init: (Command) -> Unit): CommandBuilder {
        this.init = init
        return this
    }

    fun runForTime(millis: Long): CommandBuilder {
        this.cond = { c: Command -> c.runTimeMillis() < millis }
        return this
    }

    fun withTimeout(millis: Long): CommandBuilder {
        this.timeout = millis
        return this
    }

    fun runWhile(cond: (Command) -> Boolean): CommandBuilder {
        this.cond = cond
        return this
    }

    fun onEnd(onEnd: (Command) -> Unit): CommandBuilder {
        this.onEnd = onEnd
        return this
    }

    fun onInterrupt(onInt: (Command) -> Unit): CommandBuilder {
        this.onInt = onInt
        return this
    }

    fun require(s: Subsystem): CommandBuilder {
        subsystems.add(s)
        return this
    }

    fun requires(vararg s: Subsystem): CommandBuilder {
        for (sys in s) {
            subsystems.add(sys)
        }
        return this
    }

    fun build(): Command {
        return object : Command(subsystems, timeout) {

            override val isFinished: Boolean
                get() = !cond(this)

            override fun initialize() {
                init(this)
            }

            override fun end() {
                onEnd(this)
            }

            override fun interrupted() {
                onInt(this)
            }

            override fun execute() {
                act(this)
            }
        }
    }

    companion object {
        fun create(): CommandBuilder {
            return CommandBuilder()
        }
    }
}
